In [1]:
from IPython.display import Image
from IPython.display import SVG

# <a id="1.Summary"> 1.Summary</a>

This is the solution for fetch rewards homework. Demonstrate how you reason about data and how you communicate your understanding of a specific data set to others.

-----------------------------

# <a id="2">2.Table of Contents</a>
<a href="#1.Summary">Click this Link back to Top</a>

<ol>
    <li><a href="#1.Summary">Summary</a></li>
    <li><a href="#2">Table of Contents</a></li>
    <li><a href="#3">Requirements</a></li>
    <ul>
        <li><a href="#3.1">3.1 Review unstructured JSON data and diagram a new structured relational data model</a></li>
        <ul>
            <li><a href="#3.1.1">3.1.1 receipts.json</a></li>
            <ul>
                <li><a href="#3.1.1.1">3.1.1.1 Data Preprocessing (Exploration and Observations)</a></li>
                <li><a href="#3.1.1.2">3.1.1.2 Data Modeling (Design Schema and ER Diagram)</a></li>
                <li><a href="#3.1.1.3">3.1.1.3 Data ETL (transfer JSON data to relational database)</a></li>
            </ul>
            <li><a href="#3.1.2">3.1.2 users.json</a></li>
            <ul>
                <li><a href="#3.1.2.1">3.1.2.1 Data Preprocessing (Exploration and Observations)</a></li>
                <li><a href="#3.1.2.2">3.1.2.2 Data Modeling (Design Schema and ER Diagram)</a></li>
                <li><a href="#3.1.2.3">3.1.2.3 Data ETL (transfer JSON data to relational database)</a></li>
            </ul>
            <li><a href="#3.1.3">3.1.3 brands.json</a></li>
            <ul>
                <li><a href="#3.1.3.1">3.1.3.1 Data Preprocessing (Exploration and Observations)</a></li>
                <li><a href="#3.1.3.2">3.1.3.2 Data Modeling (Design Schema and ER Diagram)</a></li>
                <li><a href="#3.1.3.3">3.1.3.3 Data ETL (transfer JSON data to relational database)</a></li>
            </ul>
            <li><a href="#3.1.4">3.1.4 Summary</a></li>
        </ul>
        <li><a href="#3.2">3.2 Write queries that directly answer predetermined questions from a business stakeholder</a></li>
        <ul>
            <li><a href="#3.2.1">3.2.1 What are the top 5 brands by receipts scanned for most recent month?</a></li>
            <li><a href="#3.2.2">3.2.2 How does the ranking of the top 5 brands by receipts scanned for the recent month compare to the ranking for the previous month?</a></li>
            <li><a href="#3.2.3">3.2.3 When considering average spend from receipts with 'rewardsReceiptStatus’ of ‘Accepted’ or ‘Rejected’, which is greater?</a></li>
            <li><a href="#3.2.4">3.2.4 When considering total number of items purchased from receipts with 'rewardsReceiptStatus’ of ‘Accepted’ or ‘Rejected’, which is greater?</a></li>
            <li><a href="#3.2.5">3.2.5 Which brand has the most spend among users who were created within the past 6 months?</a></li>
            <li><a href="#3.2.6">3.2.6 Which brand has the most transactions among users who were created within the past 6 months?</a></li>
        </ul>
        <li><a href="#3.3">3.3 Evaluate Data Quality Issues in the Data Provided</a></li>
        <li><a href="#3.4">3.4 Communicate with Stakeholders</a></li>
    </ul>
</ol>

--------------------------

# <a id="3">3.Requirements</a>
<a href="#1.Summary">Click this Link back to Top</a>

If you want to know the analysis process details, you can go through the next few secionts.
If you want to know the result, you can go to the last <a href="#3.1.4">3.1.4 Summary</a> ection.

## <a id="3.1">3.1 Review unstructured JSON data and diagram a new structured relational data model</a>
<a href="#1.Summary">Click this Link back to Top</a>

I will provide some thinking process here. Then provide the code for this data cleaning process.
1. The goal of this requirement is to clean the data from non-relational database format to relational database format for further analysis. We need to do some data cleaning process first
2. Unzip three JSON files and find these files are typicall front-end extact data. Probablay from MongoDB or other NoSQL database.
3. 

### <a id="3.1.1">3.1.1 Receipts.json</a>
<a href="#1.Summary">Click this Link back to Top</a>

Analyze receipts.json file, flatten nested JSON data and deisgn a normalized schema for relational database.

#### <a id="3.1.1.1">3.1.1.1 Data Preprocessing</a>
<a href="#1.Summary">Click this Link back to Top</a>

We open receipts.json. We find they are not standard format. Its hard to identify them.
Using JSON Crack to get a quick look at the data.
<img src = "../images/001.png">

We find receipts.json is a nested JSON data and we will failed if we directly use SQL to input this data.
The best strategy is to use MongoDB to import data first and then use other tools to transform data into relational format and database. But it is also depends on the parsing result from MongoDB.
Import receipts.json into MongoDB:
<img src = "../images/002.png">

We understand a few things from analysing the JSON schema:
1. Total records in receipts = 1119 (need to double check when you import to relational database)
2. _id: unique identifier for each records (candidate for primary key)
3. Some schema have null values, which means we need to all NULL value for this column in schema design.
4. The date format need to be transformed to target relational database date format

<img src = "../images/003.png">

We need to identify which keys are mandatory key/column and which keays are optional keys/columns.
- Export the data from MongoDB to get a standardized format and named as receipts_std.json.
- Copy receipts_std.json to JSON Crack to get the schema: receipts_std_schema.json.
- Use JSON Crack to import receipts_std_schema.json and visulizae the detailed information of the schema.
-------------------------------------------------
From the schema image below, we can get the following information:
1. The following keys can be mapped directly to columns and they are mandatory keys/columns.
    - _id
    - createDate
    - dateScanned   
    - modifyDate
    - rewardsReiptsStatus
    - userId
 2. rewardsReceiptsDetails is an array of objects. It will be a sepreated table. Only `partnerItemId` is mandatory key. This key looks like a repeatable value. So we need to use another key or keys as table primary key. 
<img src="../docs/receipts_std_schema.jpeg">

--------------

Finally, we use DBSchema to get the final relational database schema design.

|Idx |Name |Data Type | Comments |
|---|---|---|---|
| * &#128273; | \_id| objectId  | Primary Key |
|  | bonusPointsEarned| int | |
|  | bonusPointsEarnedReason| string | |
| * | createDate| date  | NOT NULL (required) |
| * | dateScanned| date  | NOT NULL (required) |
| * | modifyDate| date  | NOT NULL (required) |
|  | pointsAwardedDate| date | |
|  | pointsEarned| string  | |
|  | purchaseDate| date  | |
|  | purchasedItemCount| int | |
| * | rewardsReceiptStatus| string  | NOT NULL (required) |
|  | totalSpent| string  | |
| * | userId| string  | NOT NULL (required) |
|  | finishedDate| date | |
|  | rewardsReceiptItemList| array[object]  | Seperated Table |
|  | rewardsReceiptItemList.barcode| string  | Seperated Table |
|  | rewardsReceiptItemList.competitiveProduct| Boolean  | Seperated Table |
|  | rewardsReceiptItemList.finalPrice| string  | Seperated Table |
|  | rewardsReceiptItemList.itemPrice| string  | Seperated Table |
| * | rewardsReceiptItemList.partnerItemId| string  | Seperated Table |
|  | rewardsReceiptItemList.quantityPurchased| int  | Seperated Table |
|  | rewardsReceiptItemList.rewardsGroup| string  | Seperated Table |
|  | rewardsReceiptItemList.rewardsProductPartnerId| string  | Seperated Table |
|  | rewardsReceiptItemList.needsFetchReview| Boolean  | Seperated Table |
|  | rewardsReceiptItemList.needsFetchReviewReason| string  | Seperated Table |
|  | rewardsReceiptItemList.preventTargetGapPoints| Boolean  | Seperated Table |
|  | rewardsReceiptItemList.userFlaggedBarcode| string  | Seperated Table |
|  | rewardsReceiptItemList.userFlaggedDescription| string  | Seperated Table |
|  | rewardsReceiptItemList.userFlaggedNewItem| Boolean  | Seperated Table |
|  | rewardsReceiptItemList.userFlaggedPrice| string  | Seperated Table |
|  | rewardsReceiptItemList.userFlaggedQuantity| int  | Seperated Table |
|  | rewardsReceiptItemList.originalFinalPrice| string  | Seperated Table |
|  | rewardsReceiptItemList.originalMetaBriteBarcode| string  | Seperated Table |
|  | rewardsReceiptItemList.originalMetaBriteItemPrice| string  | Seperated Table |
|  | rewardsReceiptItemList.originalMetaBriteQuantityPurchased| int  | Seperated Table |
|  | rewardsReceiptItemList.description| string  | Seperated Table |
|  | rewardsReceiptItemList.originalMetaBriteDescription| string  | Seperated Table |
|  | rewardsReceiptItemList.pointsEarned| string  | Seperated Table |
|  | rewardsReceiptItemList.pointsPayerId| string  | Seperated Table |
|  | rewardsReceiptItemList.targetPrice| string  | Seperated Table |
|  | rewardsReceiptItemList.discountedItemPrice| string  | Seperated Table |
|  | rewardsReceiptItemList.itemNumber| string  | Seperated Table |
|  | rewardsReceiptItemList.pointsNotAwardedReason| string  | Seperated Table |
|  | rewardsReceiptItemList.brandCode| string  | Seperated Table |
|  | rewardsReceiptItemList.competitorRewardsGroup| string  | Seperated Table |
|  | rewardsReceiptItemList.originalReceiptItemText| string  | Seperated Table |
|  | rewardsReceiptItemList.deleted| Boolean  | Seperated Table |
|  | rewardsReceiptItemList.priceAfterCoupon| string  | Seperated Table |


<img src="../docs/receipts_std.svg">

--------------------------

#### <a id="3.1.1.2">3.1.1.2 Data Modeling (Design Schema and ER Diagram)</a>
<a href="#1.Summary">Click this Link back to Top</a>

For receipts.json, we need to split the data into two tables: `receipts` and `rewards_receipt_items`.
In this section, we will do the following:
1. Create a new `fetch_db` database in postgreSQL.
2. Verifty whether `_id.$oid` is unique in the `receipts` table. If so, we can use `_id.$oid` as the primary key.
3. Design the schema for `receipts` table.
4. Create `receipts` table with corresponding columns, data types, and constraints in `fetch_rewards` database.
5. Validate the schema for `receipts` table.
6. Design the schema for `rewards_receipt_items` table.
7. Create `rewards_receipt_items` table with corresponding columns, data types, and constraints in `fetch_rewards` database.
8. Validate the schema for `rewards_receipt_items` table and ER Diagram.

----------------

We choose PostgreSQL as our relational database.

<img src = "../images/13_postgresql-logo.png">

In [210]:
# %load_ext sql
%reload_ext sql
%sql postgresql://postgres:00wasabi00@155.248.207.242:5432/fetch_db

'Connected: postgres@fetch_db'

----------------

In [31]:
# 2. Verifty whether `_id.$oid` is unique in the `receipts` table. If so, we can use `_id.$oid` as the primary key.
# if `_id.$oid` is unique, it will print "True, 'All _id.$oid are unique.'"
import json

def is_oid_unique(json_data_list):
    oid_values = []
    for entry in json_data_list:
        # Check if "_id" and "$oid" exist
        if not entry.get("_id"):
            print("missing _id", entry)
            return False, "Missing '_id' in one or more entries."
        elif not entry["_id"].get("$oid"):
            print("missing oid", entry)
            return False, "Missing '$oid' in one or more entries."
        oid = entry["_id"]["$oid"]
        oid_values.append(oid)
    
    # Check uniqueness
    unique_oids = set(oid_values)
    if len(oid_values) != len(unique_oids):
        duplicates = [oid for oid in unique_oids if oid_values.count(oid) > 1]
        return False, f"Duplicate ObjectIds found: {duplicates}"
    return True, "All _id.$oid are unique."

data_list = []
data_error_list = []
with open('../data/receipts.json', 'r') as file:
    for line in file:
        try:
            data = json.loads(line)
            data_list.append(data)
        except json.JSONDecodeError as e:
            data_error_list.append(line)
            print(f"Error decoding JSON: {e}")
is_oid_unique(data_list)

(True, 'All _id.$oid are unique.')

----------------

3. Design the schema of receipts:

|Idx |JSON Name | PostgreSQL Column Name | PostgresSQL Data Type | Key Constraints | Description |
|---|---|---|---|---|---|
| * | \_id.$oid | receipt_id |  VARCHAR(24)  | Primary Key | uuid for this request |
| * | createDate.$date| create_date | TIMESTAMPTZ  | NOT NULL | The date that the event was created |
| * | dateScanned.$date| date_scanned | TIMESTAMPTZ | NOT NULL (required) | Date that the user scanned their receipt |
| * | modifyDate.$date| modify_date | TIMESTAMPTZ | NOT NULL (required) | The date the event was modified |
| * | rewardsReceiptStatus| rewards_receipt_status | VARCHAR(50) | NOT NULL (required) | status of the receipt through receipt validation and processing |
| * | userId| user_id | VARCHAR(24) | NOT NULL (required) | string id back to the User collection for the user who scanned the receipt |
|  | bonusPointsEarned| bonus_points_earned | DECIMAL(10,2) | | Number of bouns points that were awarded upon receipt completion |
|  | bonusPointsEarnedReason| bonus_points_earned_reason | TEXT  |  | event that triggered bonus points |
|  | pointsAwardedDate.$date| points_awarded_date | TIMESTAMPTZ  |  | The date we awarded points for the transaction |
|  | pointsEarned| points_earned | DECIMAL(10,2) | | The number of points earned for the receipt |
|  | purchaseDate.$date| purchase_date | TIMESTAMPTZ | | the date of the purchase |
|  | purchasedItemCount| purchased_item_count | INTEGER | | Count of number of items on the receipt |
|  | totalSpent| total_spent | DECIMAL(10,2)  | | The total amount on the receipt | 
|  | finishedDate.$date| finished_date | TIMESTAMPTZ | | Date that the receipt finished processing |


-------------

4. Create `receipts` table with corresponding columns, data types, and constraints in `fetch_rewards` database.



In [ ]:
%%sql
# DROP TABLE IF EXISTS receipts CASCADE;
-- create receipts table based on previous designed schema
-- receipt_id and user_id are typical MongoDB object id. So the best data type for them is VARCHAR(24)
-- TIMESTAMPTZ will have time zone information compared to TIMESTAMP without time zone
-- Only first 6 columns are NOT NULL from previous data exploration
-- For the data which miss in JSON, I will not provide default to handle NULL constraings value to keep the original JSON data
CREATE TABLE receipts (
    receipt_id VARCHAR(24) PRIMARY KEY,
    create_date TIMESTAMPTZ NOT NULL,
    date_scanned TIMESTAMPTZ NOT NULL,
    modify_date TIMESTAMPTZ NOT NULL,
    rewards_receipt_status VARCHAR(50) NOT NULL,
    user_id VARCHAR(24) NOT NULL,
    bonus_points_earned DECIMAL(10, 2),
    bonus_points_earned_reason TEXT,
    points_awarded_date TIMESTAMPTZ,
    points_earned DECIMAL(10, 2),
    purchase_date TIMESTAMPTZ,
    purchased_item_count INTEGER,
    total_spent DECIMAL(10, 2),
    finished_date TIMESTAMPTZ
);
SELECT * FROM receipts LIMIT 1;

 * postgresql://postgres:***@155.248.207.242:5432/fetch_db
Done.
Done.
0 rows affected.


[]

---------------------

5. Validate the schema for `receipts` table.

<img src = "../images/004.png">




------------------

6. Design the schema of `rewards_receipt_item` table:
   - `item_id` will be auto-generated as SERIAL data type by postgreSQL since we can't find a orignial data as primary key.
   - I also considered use combine columns as primary key. But the function of primary key in here is to create INDEX, so I choose to use SERIAL auto-generated data type.
   - `receipt_id` will be a foreign key that references `receipt.receipt_id` to link with `receipt` table.
   - Price related columns will use DECIMAL(10,2) data type.
   - Columns that expected have long string/text will use TEXT data type.
   - The other columns will use VARCHAR(50/255) since VARCHAR(n) don't have major performance difference with TEXT data type.
   - Total 35 columns in `rewards_receipt_item` table.

|Idx |JSON Name | PostgreSQL Column Name | PostgresSQL Data Type | Key Constraints |
|---|---|---|---|---|
| 1 |  | item_id | SERIAL | Primary Key |
| 2 | _is.$oid | receipt_id | VARCHAR(24) | Foreign Key (references receipt.receipt_id) |
| 3 | rewardsReceiptItemList.partnerItemId| partener_item_id | VARCHAR(24) | NOT NULL |
| 4 | rewardsReceiptItemList.barcode| barcode | VARCHAR(25) | |
| 5 | rewardsReceiptItemList.competitiveProduct | competitive_product | BOOLEAN | NOT NULL (required) |
| 6 | rewardsReceiptItemList.finalPrice| final_price | DECIMAL(10,2) |  |
| 7 | rewardsReceiptItemList.itemPrice | item_price | DECIMAL(10,2) | |
| 8 | rewardsReceiptItemList.quantityPurchased | quantity_purchased | INTEGER |  |
| 9 | rewardsReceiptItemList.rewardsGroup | rewards_group | VARCHAR(100) |  |
| 10 | rewardsReceiptItemList.rewardsProductPartnerId | rewards_product_partner_id | VARCHAR(50) |  |
| 11 | rewardsReceiptItemList.needsFetchReview | needs_fetch_review | BOOLEAN |  |
| 12 | rewardsReceiptItemList.needsFetchReviewReason | needs_fetch_review_reason | TEXT |  |
| 13 | rewardsReceiptItemList.preventTargetGapPoints | prevent_target_gap_points | BOOLEAN |  |
| 14 | rewardsReceiptItemList.userFlaggedBarcode | user_flagged_barcode | VARCHAR(255) |  |
| 15 | rewardsReceiptItemList.userFlaggedDescription | user_flagged_description | TEXT |  |
| 16 | rewardsReceiptItemList.userFlaggedNewItem | user_flagged_new_item | BOOLEAN |  |
| 17 | rewardsReceiptItemList.userFlaggedPrice | user_flagged_price | DECIMAL(10,2) |  |
| 18 | rewardsReceiptItemList.userFlaggedQuantity | user_flagged_quantity | INTEGER |  |
| 19 | rewardsReceiptItemList.originalFinalPrice | original_final_price | DECIMAL(10,2) |  |
| 20 | rewardsReceiptItemList.originalMetaBriteBarcode | original_meta_brite_barcode | VARCHAR(255) |  |
| 21 | rewardsReceiptItemList.originalMetaBriteItemPrice | original_meta_brite_item_price | DECIMAL(10,2) |  |
| 22 | rewardsReceiptItemList.originalMetaBriteQuantityPurchased | original_meta_brite_quantity_purchased | INTEGER |  |
| 23 | rewardsReceiptItemList.description | description | TEXT |  |
| 24 | rewardsReceiptItemList.originalMetaBriteDescription | original_meta_brite_description | TEXT |  |
| 25 | rewardsReceiptItemList.pointsEarned | points_earned | VARCHAR(50) |  |
| 26 | rewardsReceiptItemList.pointsPayerId | points_payer_id | VARCHAR(50) |  |
| 27 | rewardsReceiptItemList.targetPrice | target_price | DECIMAL(10,2) |  |
| 28 | rewardsReceiptItemList.discountedItemPrice | discounted_item_price | DECIMAL(10,2) |  |
| 29 | rewardsReceiptItemList.itemNumber | item_number | VARCHAR(50) |  |
| 30 | rewardsReceiptItemList.pointsNotAwardedReason | points_not_awarded_reason | TEXT |  |
| 31 | rewardsReceiptItemList.brandCode | brand_code | VARCHAR(50) |  |
| 32 | rewardsReceiptItemList.competitorRewardsGroup | competitor_rewards_group | VARCHAR(100) |  |
| 33 | rewardsReceiptItemList.originalReceiptItemText | original_receipt_item_text | TEXT |  |
| 34 | rewardsReceiptItemList.deleted | deleted | BOOLEAN |  |
| 35 | rewardsReceiptItemList.priceAfterCoupon | price_after_coupon | DECIMAL(10,2) |  |

-----------------

7. Create `rewards_receipt_items` table with corresponding columns, data types, and constraints in `fetch_rewards` database.

In [97]:
%%sql
-- add DROP sql to quickly reset table
DROP TABLE IF EXISTS rewards_receipt_items;

CREATE TABLE IF NOT EXISTS rewards_receipt_items (
    item_id SERIAL PRIMARY KEY,
    -- foreign key to Parent receipts table
    receipt_id VARCHAR(24) NOT NULL REFERENCES receipts(receipt_id),
    partner_item_id VARCHAR(24) NOT NULL,
    barcode VARCHAR(25),
    competitive_product BOOLEAN,
    final_price VARCHAR(24),
    item_price DECIMAL(10,2),
    quantity_purchased INTEGER,
    rewards_group VARCHAR(100),
    rewards_product_partner_id VARCHAR(24),
    needs_fetch_review BOOLEAN,
    needs_fetch_review_reason TEXT,
    prevent_target_gap_points BOOLEAN,
    user_flagged_barcode VARCHAR(255),
    user_flagged_description TEXT,
    user_flagged_new_item BOOLEAN,
    user_flagged_price DECIMAL(10,2),
    user_flagged_quantity INTEGER,
    original_final_price DECIMAL(10,2),
    original_meta_brite_barcode VARCHAR(255),
    original_meta_brite_item_price DECIMAL(10,2),
    original_meta_brite_quantity_purchased INTEGER,
    description TEXT,
    original_meta_brite_description TEXT,
    points_earned VARCHAR(50),
    points_payer_id VARCHAR(50),
    target_price DECIMAL(10,2),
    discounted_item_price DECIMAL(10,2),
    -- what is this? id or count?
    item_number VARCHAR(50),
    points_not_awarded_reason TEXT,
    brand_code VARCHAR(50),
    competitor_rewards_group VARCHAR(100),
    original_receipt_item_text TEXT,
    -- only 9 records have this field populated
    deleted BOOLEAN,
    price_after_coupon DECIMAL(10,2)
);

SELECT * FROM rewards_receipt_items;
SELECT COUNT(*)
FROM information_schema.columns
WHERE table_name = 'rewards_receipt_items'

 * postgresql://postgres:***@155.248.207.242:5432/fetch_db
Done.
Done.
0 rows affected.
1 rows affected.


[(35,)]

-----------

8. Validate the schema for `rewards_receipt_items` table and ER Diagram (powered by DB Schema)

<img src = "../images/005.png">

##### Summary
1. I created two tables `receipts` and `rewards_receipt_items` with designed schema.
2. 

<br>

-------------------------------------

<br>

#### <a id="3.1.1.3">3.1.1.3 Data ETL (transfer JSON data to relational database)</a>
<a href="#1.Summary">Click this Link back to Top</a>

I choose to use Solution 1 SQL as the major tool for data ETL.
Prons: easy to use
Cons: limited flexibility and don't have erro handling and logging capability.

| Idx | &#9989; Solution 1: Use SQL to INSERT data | Solution 2: Use Python to IMPORT data |
|---|---|---|
| 1 | Create a temporary table to store JSON raw data | Use Python library pyscog2 or asyncpg to INSERT data into target table |
| 2 | Use python LOAD raw JSON into JSONB type column |  |
| 3 | Use SQL to extract data from JSONB column and Flatten & INSERT into `receipts` table |  |
| 4 | Use SQL to extract data from JSONB column and Flatten & INSERT into `rewards_receipt_items`  table|  |
| 5 | Validate two tables import successfully| | 

1. Create a temporary table to store JSON raw data 

In [71]:
%%sql
CREATE TABLE receipts_raw (
    id SERIAL PRIMARY KEY,
    data JSONB
);
SELECT * FROM receipts_raw;

 * postgresql://postgres:***@155.248.207.242:5432/fetch_db
(psycopg2.errors.DuplicateTable) relation "receipts_raw" already exists

[SQL: CREATE TABLE receipts_raw (
    id SERIAL PRIMARY KEY,
    data JSONB
);]
(Background on this error at: http://sqlalche.me/e/14/f405)


-----------------------

2. Use python LOAD raw JSON into JSONB type column

In [55]:
import psycopg2
import json

# Connect to the remote PostgreSQL database
# [TODO]: refactor the code by using asynchio to handle large datasets
# [TODO]: replace the connection information with .env related variables to hide sensitive information
db_connection = psycopg2.connect(
    host="155.248.207.242",   
    dbname="fetch_db",
    user="postgres",
    password="00wasabi00",
    port="5432"
)
cursor = db_connection.cursor()


data_list = []
data_error_list = []
with open('../data/receipts.json', 'r') as file:
    for line in file:
        try:
            data = json.loads(line)
            data_list.append(data)
        except json.JSONDecodeError as e:
            data_error_list.append(line)
            print(f"Error decoding JSON: {e}")

try:
    # Insert each JSON object separately
    insert_query = "INSERT INTO receipts_raw (data) VALUES (%s)"

    for record in data_list:
        cursor.execute(insert_query, (json.dumps(record),))  # Convert dict to JSONB format
    
    # Commit changes
    db_connection.commit()
    print(f"Successfully inserted {len(data_list)} records into PostgreSQL!")

except Exception as e:
    print(f"Error: {e}")

finally:
    # Close the connection
    cursor.close()
    db_connection.close()



✅ Successfully inserted 1119 records into PostgreSQL!


In [74]:
%%sql
SELECT data FROM receipts_raw LIMIT 1;

 * postgresql://postgres:***@155.248.207.242:5432/fetch_db
1 rows affected.


[({'_id': {'$oid': '5ff1e1eb0a720f0523000575'}, 'userId': '5ff1e1eacfcf6c399c274ae6', 'createDate': {'$date': 1609687531000}, 'modifyDate': {'$date': 16 ... (629 characters truncated) ... 5, 'preventTargetGapPoints': True}], 'bonusPointsEarnedReason': 'Receipt number 2 completed, bonus point schedule DEFAULT (5cefdcacf3693e0b50e83a36)'},)]

----------------

3. Use SQL to extract data from JSONB column and Flatten & INSERT into 'receipts'table

In [185]:
%%sql
INSERT INTO receipts (
    receipt_id,
    create_date,
    date_scanned,
    modify_date,
    rewards_receipt_status,
    user_id,
    bonus_points_earned,
    bonus_points_earned_reason,
    points_awarded_date,
    points_earned,
    purchase_date,
    purchased_item_count,
    total_spent,
    finished_date
)
SELECT
    (data -> '_id' ->> '$oid')::VARCHAR(24) AS receipt_id,
    -- MongoDB stores dates in millsecons, convert Milliseconds to Seconds
    TO_TIMESTAMP((data -> 'createDate' ->> '$date')::BIGINT / 1000) AS create_date,
    TO_TIMESTAMP((data -> 'dateScanned' ->> '$date')::BIGINT / 1000) AS date_scanned,
    TO_TIMESTAMP((data -> 'modifyDate' ->> '$date')::BIGINT / 1000) AS modify_date,
    (data ->> 'rewardsReceiptStatus')::VARCHAR(50) AS rewards_receipt_status,
    (data ->> 'userId')::VARCHAR(24) AS user_id,
    (data ->> 'bonusPointsEarned')::DECIMAL(10,2) AS bonus_points_earned,
    (data ->> 'bonusPointsEarnedReason')::TEXT AS bonus_points_earned_reason,
    TO_TIMESTAMP((data -> 'pointsAwardedDate' ->> '$date')::BIGINT / 1000) AS points_awarded_date,
    (data ->> 'pointsEarned')::DECIMAL(10,2) AS points_earned,
    TO_TIMESTAMP((data -> 'purchaseDate' ->> '$date')::BIGINT / 1000) AS purchase_date,
    (data ->> 'purchasedItemCount')::INTEGER AS purchased_item_count,
    (data ->> 'totalSpent')::DECIMAL(10,2) AS total_spent,
    TO_TIMESTAMP((data -> 'finishedDate' ->> '$date')::BIGINT / 1000) AS finished_date
FROM 
    receipts_raw;

SELECT * FROM receipts LIMIT 1;

 * postgresql://postgres:***@155.248.207.242:5432/fetch_db
1119 rows affected.
1 rows affected.


[('5ff1e1eb0a720f0523000575', datetime.datetime(2021, 1, 3, 15, 25, 31, tzinfo=datetime.timezone.utc), datetime.datetime(2021, 1, 3, 15, 25, 31, tzinfo=datetime.timezone.utc), datetime.datetime(2021, 1, 3, 15, 25, 36, tzinfo=datetime.timezone.utc), 'FINISHED', '5ff1e1eacfcf6c399c274ae6', Decimal('500.00'), 'Receipt number 2 completed, bonus point schedule DEFAULT (5cefdcacf3693e0b50e83a36)', datetime.datetime(2021, 1, 3, 15, 25, 31, tzinfo=datetime.timezone.utc), Decimal('500.00'), datetime.datetime(2021, 1, 3, 0, 0, tzinfo=datetime.timezone.utc), 5, Decimal('26.00'), datetime.datetime(2021, 1, 3, 15, 25, 31, tzinfo=datetime.timezone.utc))]

--------------------------------

4. Use SQL to extract data from JSONB column and Flatten & INSERT into `rewards_receipt_items`  table

In [99]:
%%sql
INSERT INTO rewards_receipt_items (
    receipt_id,
    partner_item_id,
    barcode,
    competitive_product,
    final_price,
    item_price,
    quantity_purchased,
    rewards_group,
    rewards_product_partner_id,
    needs_fetch_review,
    needs_fetch_review_reason,
    prevent_target_gap_points,
    user_flagged_barcode,
    user_flagged_description,
    user_flagged_new_item,
    user_flagged_price,
    user_flagged_quantity,
    original_final_price,
    original_meta_brite_barcode,
    original_meta_brite_item_price,
    original_meta_brite_quantity_purchased,
    description,
    original_meta_brite_description,
    points_earned,
    points_payer_id,
    target_price,
    discounted_item_price,
    item_number,
    points_not_awarded_reason,
    brand_code,
    competitor_rewards_group,
    original_receipt_item_text,
    deleted,
    price_after_coupon
)
SELECT
    (data -> '_id' ->> '$oid')::VARCHAR(24) AS receipt_id,
    (item ->> 'partnerItemId')::VARCHAR(24) AS partner_item_id,
    (item ->> 'barcode')::VARCHAR(25) AS barcode,
    (item ->> 'competitiveProduct')::BOOLEAN AS competitive_product,
    (item ->> 'finalPrice')::DECIMAL(10,2) AS final_price,
    (item ->> 'itemPrice')::DECIMAL(10,2) AS item_price,
    (item ->> 'quantityPurchased')::INTEGER AS quantity_purchased,
    (item ->> 'rewardsGroup')::VARCHAR(100) AS rewards_group,
    (item ->> 'rewardsProductPartnerId')::VARCHAR(50) AS rewards_product_partner_id,
    (item ->> 'needsFetchReview')::BOOLEAN AS needs_fetch_review,
    (item ->> 'needsFetchReviewReason')::TEXT AS needs_fetch_review_reason,
    (item ->> 'preventTargetGapPoints')::BOOLEAN AS prevent_target_gap_points,
    (item ->> 'userFlaggedBarcode')::VARCHAR(255) AS user_flagged_barcode,
    (item ->> 'userFlaggedDescription')::TEXT AS user_flagged_description,
    (item ->> 'userFlaggedNewItem')::BOOLEAN AS user_flagged_new_item,
    (item ->> 'userFlaggedPrice')::DECIMAL(10,2) AS user_flagged_price,
    (item ->> 'userFlaggedQuantity')::INTEGER AS user_flagged_quantity,
    (item ->> 'originalFinalPrice')::DECIMAL(10,2) AS original_final_price,
    (item ->> 'originalMetaBriteBarcode')::VARCHAR(255) AS original_meta_brite_barcode,
    (item ->> 'originalMetaBriteItemPrice')::DECIMAL(10,2) AS original_meta_brite_item_price,
    (item ->> 'originalMetaBriteQuantityPurchased')::INTEGER AS original_meta_brite_quantity_purchased,
    (item ->> 'description')::TEXT AS description,
    (item ->> 'originalMetaBriteDescription')::TEXT AS original_meta_brite_description,
    (item ->> 'pointsEarned')::VARCHAR(50) AS points_earned,
    (item ->> 'pointsPayerId')::VARCHAR(50) AS points_payer_id,
    (item ->> 'targetPrice')::DECIMAL(10,2) AS target_price,
    (item ->> 'discountedItemPrice')::DECIMAL(10,2) AS discounted_item_price,
    (item ->> 'itemNumber')::VARCHAR(50) AS item_number,
    (item ->> 'pointsNotAwardedReason')::TEXT AS points_not_awarded_reason,
    (item ->> 'brandCode')::VARCHAR(50) AS brand_code,
    (item ->> 'competitorRewardsGroup')::VARCHAR(100) AS competitor_rewards_group,
    (item ->> 'originalReceiptItemText')::TEXT AS original_receipt_item_text,
    (item ->> 'deleted')::BOOLEAN AS deleted,
    (item ->> 'priceAfterCoupon')::DECIMAL(10,2) AS price_after_coupon
FROM 
    receipts_raw,
LATERAL 
    jsonb_array_elements(data -> 'rewardsReceiptItemList') AS item;

SELECT COUNT(*) FROM rewards_receipt_items;

 * postgresql://postgres:***@155.248.207.242:5432/fetch_db
6941 rows affected.
1 rows affected.


[(6941,)]

In [100]:
%%sql
SELECT * FROM rewards_receipt_items LIMIT 1;

 * postgresql://postgres:***@155.248.207.242:5432/fetch_db
1 rows affected.


[(1, '5ff1e1eb0a720f0523000575', '1', '4011', None, '26.00', Decimal('26.00'), 5, None, None, False, None, True, '4011', None, True, Decimal('26.00'), 5, None, None, None, None, 'ITEM NOT FOUND', None, None, None, None, None, None, None, None, None, None, None, None)]

-----------------

5. Validate two tables import successfully and summary
   I created two tables, `receipts` and `rewards_receipt_items`.

In [106]:
%%sql
-- join to validate design and INSERT are successful
SELECT
    r.receipt_id,
    r.create_date,
    r.date_scanned,
    r.user_id,
    ri.item_id,
    ri.partner_item_id,
    ri.barcode,
    ri.item_price
FROM
    receipts r
    INNER JOIN rewards_receipt_items ri ON r.receipt_id = ri.receipt_id
LIMIT 3;

 * postgresql://postgres:***@155.248.207.242:5432/fetch_db
3 rows affected.


[('5ff1e1eb0a720f0523000575', datetime.datetime(2021, 1, 3, 15, 25, 31, tzinfo=datetime.timezone.utc), datetime.datetime(2021, 1, 3, 15, 25, 31, tzinfo=datetime.timezone.utc), '5ff1e1eacfcf6c399c274ae6', 1, '1', '4011', Decimal('26.00')),
 ('5ff1e1bb0a720f052300056b', datetime.datetime(2021, 1, 3, 15, 24, 43, tzinfo=datetime.timezone.utc), datetime.datetime(2021, 1, 3, 15, 24, 43, tzinfo=datetime.timezone.utc), '5ff1e194b6a9d73a3a9f1052', 2, '1', '4011', Decimal('1.00')),
 ('5ff1e1bb0a720f052300056b', datetime.datetime(2021, 1, 3, 15, 24, 43, tzinfo=datetime.timezone.utc), datetime.datetime(2021, 1, 3, 15, 24, 43, tzinfo=datetime.timezone.utc), '5ff1e194b6a9d73a3a9f1052', 3, '2', '028400642255', Decimal('10.00'))]

----------------------------------

<br>

### <a id="3.1.2">3.1.2 users.json</a>

<a href="#1.Summary">Click this Link back to Top</a>

Analyze users.json and design a relational normalized schema to store the data.

-------------------------

#### <a id="3.1.2.1">3.1.2.1 Data Preprocessing (Exploration and Observation)</a>

<a href="#1.Summary">Click this Link back to Top</a>

Import into MongoDB. We find a lot of duplicate data in this dataset when you consider `id.$oid` as the unique identifier. 
If we regard `id.$oid` as user_id, it will make sense. But in the meantime, we need to create a new column/filed as the unique identifier pramay key.
1. `id.$oid` is not a unique identifier. We need to create a new column/filed as the unique identifier pramay key.
2. `active` , `createdDate`, and `role` are required and NOT NULL columns/fileds. 
3. `lastLogin`, `signUpScource`, and `state` have missing value in the collecion.
4. We import total 212 records into MongoDB but there a lot of duplicate data need to be handled correctly.

<img src = "../images/007.png">

Visulization
1. Export JSON file from mongoDB to get a standardized format JSON for visualization.
2. Use JSON Crack Schema Validator & Generator to get a standardized JSON schema. Restore as `users_schema.json` in local directory.
3. Use JSON Crack Extension to get visulization.
   
<img src="../docs/users_std_schema.jpeg">

Finally, we use DBSchema to get the final schema for users.json file.
We need to auto-generated a new unique ID for postgreSQL.

| Column         | Type       | Constraints           | Description                            |
|---------------|-----------|-----------------------|------------------------------------------|
| _id.$oid     | TEXT    | duplicated                | user Id   |
| active      | BOOLEAN | NOT NULL                     |  indicates if the user is active; only Fetch will de-activate an account with this flag  |
| createdDate.$date| TIMESTAMP |                       | when the user created their account |
| lastLogin.$date  | TIMESTAMP | NULLABLE          | last time the user was recorded logging in to the app |
| role        | TEXT    |                           | Sconstant value set to 'CONSUMER'|
| signUpSource | TEXT | NULLABLE                    | Stores the sign-up source.              |
| state`      | TEXT    | NULLABLE                   | state abbreviation                  |

<img src="../images/008.png">

---------------------------

#### <a id="3.1.2.2">3.1.2.2 Data Modeling (Design Schema and ER Diagram)</a>

<a href="#1.Summary">Click this Link back to Top</a>

1. Verify whether `_id.$oid` is unique or not. If not, we need a new field as the PostgreSQL primary key.
2. Design the schema for `users` table in PostgreSQL.
3. Create `users` table with cooresponding columns name, data type and key constrains.
4. Validate the scheam for `uers` table.

In [109]:
# 1. Verifty whether `_id.$oid` is unique in the `users` table. If so, we can use `_id.$oid` as the primary key.
# if `_id.$oid` is unique, it will print "True, 'All _id.$oid are unique.'"
import json

def is_oid_unique(json_data_list):
    oid_values = []
    for entry in json_data_list:
        # Check if "_id" and "$oid" exist
        if not entry.get("_id"):
            print("missing _id", entry)
            return False, "Missing '_id' in one or more entries."
        elif not entry["_id"].get("$oid"):
            print("missing oid", entry)
            return False, "Missing '$oid' in one or more entries."
        oid = entry["_id"]["$oid"]
        oid_values.append(oid)
    
    # Check uniqueness
    unique_oids = set(oid_values)
    if len(oid_values) != len(unique_oids):
        duplicates = [oid for oid in unique_oids if oid_values.count(oid) > 1]
        return False, f"Duplicate ObjectIds found: {duplicates}"
    return True, "All _id.$oid are unique."

data_list = []
data_error_list = []
with open('../data/users.json', 'r') as file:
    for line in file:
        try:
            data = json.loads(line)
            data_list.append(data)
        except json.JSONDecodeError as e:
            data_error_list.append(line)
            print(f"Error decoding JSON: {e}")
is_oid_unique(data_list)

(False,
 "Duplicate ObjectIds found: ['5fff0f4fb3348b03eb45abb0', '5ff5d15aeb7c7d12096d91a2', '5ffcb4bc04929111f6e92608', '5fa41775898c7a11a6bcef3e', '5ff8da28b3348b11c9337ac6', '600741d06e6469120a787853', '59c124bae4b0299e55b0f330', '600987d77d983a11f63cfa92', '600f00d05edb787dce05fb84', '5ff873d1b3348b11c9337716', '5ff370c562fde912123a5e0e', '6000d46cfb296c121a81b20c', '6000b7aefb296c121a8198b1', '5ff7264e8f142f11dd189504', '60025c65fb296c4ef805d9e6', '600ed42e43298911ce45d1fa', '60189c94c8b50e11d8454f6b', '60145a3c84231211ce796c5d', '600fb1ac73c60b12049027bb', '6007464b6e64691717e8c1f0', '60183090c8b50e11d84549c0', '5fff4beedf9ace121f0c17ea', '6008893b633aab121bb8e0a9', '600f47f06fd0dc1768a34a12', '60229990b57b8a12187fe9e0', '5ff4ce3dc3d63511e2a484dc', '5ff36a3862fde912123a4460', '600f41b2bd196811e68ea219', '5ff36be7135e7011bcb856d3', '60088d55633aab121bb8e41a', '5ff47392c3d63511e2a47881', '5ffc9d87b3348b11c9338920', '6000b75bbe5fc96dfee1d4d3', '5ffc8ff9b3348b11c9338896', '5a43c08fe

---------------

2. Design the schema for `users` table in PostgreSQL.
 - Create `record_id` UUID as the primary key for each user activity record.
 - `_id.$oid` is actually `user_id` in `receipts` table. Should be the Foreign Key.

|   | JSON Name  | PostgreSQL Column Name |Type     | Constraints           | Description                            |
|---|------------|------------------------|---------|--------------|------------------------------------------|
| 1 |            | record_id | UUID | Primary Key | Random unique identifer for each user activity record |
| 2 | _id.$oid   | user_id   | VARCHAR(24)    |  Foreign Key   | Consistent with `receipts.user_id`     |
| 3 | active      | active   | BOOLEAN | NOT NULL            | indicates if the user is active; only Fetch will de-activate an account with this flag  |
| 4 | createdDate.$date| created_date | TIMESTAMP |           | when the user created their account |
| 5 | role        | role      | TEXT    |             | constant value set to 'CONSUMER'  |
| 6 | lastLogin.$date  | last_login_date | TIMESTAMP | NULLABLE         |  last time the user was recorded logging in to the app  |
| 7 | signUpSource | sign_up_source | TEXT | NULLABLE           | Stores the sign-up source.              |
| 8 | state      | state | TEXT    | NULLABLE          | state abbreviation   |
    

-----------------------

3. Create `users` table with cooresponding columns name, data type and key constrains.

In [138]:
%%sql
DROP TABLE IF EXISTS users CASCADE;
-- create users table
CREATE TABLE IF NOT EXISTS users(
    record_id UUID DEFAULT gen_random_uuid() PRIMARY KEY, 
    user_id VARCHAR(24),
    active BOOLEAN NOT NULL,
    created_date TIMESTAMP,
    role TEXT,
    last_login_date TIMESTAMP,
    sign_up_source TEXT,
    state TEXT
);

SELECT * FROM users LIMIT 1;

 * postgresql://postgres:***@155.248.207.242:5432/fetch_db
Done.
Done.
0 rows affected.


[]

----------

4. Validate the scheam for `uers` table.
   
<img src = "../images/009.png">

------------------------

<br>

#### <a id="3.1.2.3">3.1.2.3 Data ETL (Transfer JSON data to relational database)</a>

<a href="#1.Summary">Click this Link back to Top</a>

1. Create a temporary `users_raw`table to upload JSON file into PostgreSQL database.
2. Use Pyothon to upload JSON file into the temporary `users_raw` table.
3. Use SQL to extract data from the temporary `users_raw` table and load it into a new table `users` in the PostgreSQL database.
4. Validate the `users` table by join it with `receipts` table.

------------------------------

1. Create a temporary `users_raw`table to upload JSON file into PostgreSQL database.

In [ ]:
%%sql
-- DROP TABLE IF EXISTS users_raw;
CREATE TABLE users_raw (
    id SERIAL PRIMARY KEY,
    json_data JSONB
);

SELECT * FROM users_raw;

 * postgresql://postgres:***@155.248.207.242:5432/fetch_db
Done.
Done.
0 rows affected.


[]

----------------------------------

2. Use Pyothon to upload JSON file into the temporary `users_raw` table.

In [114]:
import psycopg2
import json

# Connect to the remote PostgreSQL database
# [TODO]: refactor the code by using asynchio to handle large datasets
# [TODO]: replace the connection information with .env related variables to hide sensitive information
db_connection = psycopg2.connect(
    host="155.248.207.242",   
    dbname="fetch_db",
    user="postgres",
    password="00wasabi00",
    port="5432"
)
cursor = db_connection.cursor()


data_list = []
data_error_list = []
with open('../data/users.json', 'r') as file:
    for line in file:
        try:
            data = json.loads(line)
            data_list.append(data)
        except json.JSONDecodeError as e:
            data_error_list.append(line)
            print(f"Error decoding JSON: {e}")

try:
    # Insert each JSON object separately
    insert_query = "INSERT INTO users_raw (json_data) VALUES (%s)"

    for record in data_list:
        cursor.execute(insert_query, (json.dumps(record),))  # Convert dict to JSONB format
    
    # Commit changes
    db_connection.commit()
    print(f"Successfully inserted {len(data_list)} records into PostgreSQL!")

except Exception as e:
    print(f"Error: {e}")

finally:
    # Close the connection
    cursor.close()
    db_connection.close()

Successfully inserted 495 records into PostgreSQL!


In [117]:
%%sql
SELECT * FROM users_raw LIMIT 5;

 * postgresql://postgres:***@155.248.207.242:5432/fetch_db
5 rows affected.


[(1, {'_id': {'$oid': '5ff1e194b6a9d73a3a9f1052'}, 'role': 'consumer', 'state': 'WI', 'active': True, 'lastLogin': {'$date': 1609687537858}, 'createdDate': {'$date': 1609687444800}, 'signUpSource': 'Email'}),
 (2, {'_id': {'$oid': '5ff1e194b6a9d73a3a9f1052'}, 'role': 'consumer', 'state': 'WI', 'active': True, 'lastLogin': {'$date': 1609687537858}, 'createdDate': {'$date': 1609687444800}, 'signUpSource': 'Email'}),
 (3, {'_id': {'$oid': '5ff1e194b6a9d73a3a9f1052'}, 'role': 'consumer', 'state': 'WI', 'active': True, 'lastLogin': {'$date': 1609687537858}, 'createdDate': {'$date': 1609687444800}, 'signUpSource': 'Email'}),
 (4, {'_id': {'$oid': '5ff1e1eacfcf6c399c274ae6'}, 'role': 'consumer', 'state': 'WI', 'active': True, 'lastLogin': {'$date': 1609687530597}, 'createdDate': {'$date': 1609687530554}, 'signUpSource': 'Email'}),
 (5, {'_id': {'$oid': '5ff1e194b6a9d73a3a9f1052'}, 'role': 'consumer', 'state': 'WI', 'active': True, 'lastLogin': {'$date': 1609687537858}, 'createdDate': {'$date': 1609687444800}, 'signUpSource': 'Email'})]

--------------------------

3. Use SQL to extract data from the temporary `users_raw` table and load it into a new table `users` in the PostgreSQL database.

In [139]:
%%sql
SELECT * FROM users LIMIT 1;

 * postgresql://postgres:***@155.248.207.242:5432/fetch_db
0 rows affected.


[]

In [141]:
%%sql
-- no need to INSERT anything to record_id. It will trigger DEFAULT gen_random_uuid()
INSERT INTO users (
    user_id, 
    active, 
    created_date, 
    role, 
    last_login_date, 
    sign_up_source, 
    state
)
SELECT 
    json_data->'_id'->>'$oid' AS user_id,
    (json_data->>'active')::BOOLEAN AS active,
    TO_TIMESTAMP((json_data -> 'createdDate' ->> '$date')::BIGINT / 1000) AS created_date,
    json_data->>'role' AS role,
    TO_TIMESTAMP((json_data -> 'lastLogin' ->> '$date')::BIGINT / 1000) AS last_login_date,
    json_data->>'signUpSource' AS sign_up_source,
    json_data->>'state' AS state
FROM
    users_raw;

SELECT COUNT(*) FROM users;

 * postgresql://postgres:***@155.248.207.242:5432/fetch_db
495 rows affected.
1 rows affected.


[(495,)]

---------------

4. Validate the `users` table by join it with `receipts` table.

In [151]:
%%sql
SELECT
    u.record_id,
    u.user_id,
    r.receipt_id,
    r.create_date
FROM
    users AS u
    INNER JOIN receipts AS r ON u.user_id = r.user_id
LIMIT 4;

 * postgresql://postgres:***@155.248.207.242:5432/fetch_db
4 rows affected.


[(UUID('ed42f7d5-15ab-4493-a5a6-0532eab527ba'), '5ff1e1eacfcf6c399c274ae6', '5ff1e1eb0a720f0523000575', datetime.datetime(2021, 1, 3, 15, 25, 31, tzinfo=datetime.timezone.utc)),
 (UUID('bc9f2bb7-54d4-416d-9d94-22dec550faf3'), '5ff1e1eacfcf6c399c274ae6', '5ff1e1eb0a720f0523000575', datetime.datetime(2021, 1, 3, 15, 25, 31, tzinfo=datetime.timezone.utc)),
 (UUID('9063add1-5ffb-460e-843f-8ec8551a623f'), '5ff1e1eacfcf6c399c274ae6', '5ff1e1eb0a720f0523000575', datetime.datetime(2021, 1, 3, 15, 25, 31, tzinfo=datetime.timezone.utc)),
 (UUID('571f34e1-948a-47bf-952b-da0cd4a337a9'), '5ff1e1eacfcf6c399c274ae6', '5ff1e1eb0a720f0523000575', datetime.datetime(2021, 1, 3, 15, 25, 31, tzinfo=datetime.timezone.utc))]

<br>

-------------------

### <a id="3.1.3">3.1.3 brands.json</a>
<a href="#1.Summary">Click this Link back to Top</a>

Anaylze brands.json and design a relation normalized schema for the data.

------------------------------

#### <a id="3.1.3.1">3.1.3.1 Data Preprocessing (Exploration and Observation)</a>

<a href="#1.Summary">Click this Link back to Top</a>

Import JSON into MongoDB and analyze the schema. We found the following thing:
1. Total 1167 records.
2. `_id` and `barcode` are required fields and NOT NULL.
3. `cpg` is nested field and contains `$id.$oid`. Although `cpg` can be seperated table but each record only have one 'cpg' field. So it can also be considered as two fields `cpg.$id.$oid` and `cpg.$ref`.
4. When I tried to find the foreign key, I found `users.brandCode` is the only foreign key candidate which refers to `rewards_recipt_item.brandCode`. And I also I found `users.name` is idenitfy with `users.brandCode` most of the time. The exceiptional case is  `users.brandCode` is null and `users.name` and `users.brandCode` use the different upper and lower cases.

<img src="../images/010.png">

Visualization
1. Export JSON file from MongoDB and save it as a standardized format JSON `users_std.json` for visualization.
2. Use JSON Crack Schema Validator & Generator to get a standardized JSON Schema. Save as `users_std_schema.json` in local directory.
3. Use JSON Crack Extenstion to get visualization

<img src="../docs/brands_std_schema.jpeg">

Finally, we use DBSchema to get the mongoDB Schema for brands.json

| Column Name  | Data Type    | Key Constraints                              | Description                          |
|----------------|-------------|------------------------------------------|--------------------------------------|
| _id.$oid        | TEXT        | VARCHAR(24) | PRIMARY KEY                | brand uuid       |
| barcode          | TEXT        | NOT NULL                                | the barcode on the item    |
| name              | TEXT        | NOT NULL                                | Brand name  |
| cpg.$id.$oid     | TEXT        | NOT NULL                                |     Unique for each record     |
| cpg.$ref      | VARCHAR(24) | NOT NULL                                 |  Customer Packages Goods   reference to CPG collection       |
| categoryCode | TEXT      | NULLABLE                                        |  The category name for which the brand sells products in|
| category    | TEXT        | NULLABLE                                | The category name for which the brand sells products in |
| topBrand   | BOOLEAN     |                                           | Boolean indicator for whether the brand should be featured as a 'top brand'  |
| brandCode  | TEXT        | NULLABLE                                | String that corresponds with the brand column in a partner product file |

<br>
<img src="../images/011.png">

<br>

---------------------------

#### <a id="3.1.3.2">3.1.3.2 Data Modeling (Design Scheam)</a>

<a href="#1.Summary">Click this Link back to Top</a>

Here is the data modeling processing steps:
1. Verify whether `_id.$oid` is unique. If yes, it will be used as the primary key.
2. Design the postgreSQL schema for `brands` table.
3. Create the `brands` table by SQL with corresponding columns and data types.
4. Validate the `brands` table schema and ER-Diagram.

In [152]:
# 1. Verifty whether `_id.$oid` is unique in the `brands` JSON. If so, we can use `_id.$oid` as the primary key.
# if `_id.$oid` is unique, it will print "True, 'All _id.$oid are unique.'"
import json

def is_oid_unique(json_data_list):
    oid_values = []
    for entry in json_data_list:
        # Check if "_id" and "$oid" exist
        if not entry.get("_id"):
            print("missing _id", entry)
            return False, "Missing '_id' in one or more entries."
        elif not entry["_id"].get("$oid"):
            print("missing oid", entry)
            return False, "Missing '$oid' in one or more entries."
        oid = entry["_id"]["$oid"]
        oid_values.append(oid)
    
    # Check uniqueness
    unique_oids = set(oid_values)
    if len(oid_values) != len(unique_oids):
        duplicates = [oid for oid in unique_oids if oid_values.count(oid) > 1]
        return False, f"Duplicate ObjectIds found: {duplicates}"
    return True, "All _id.$oid are unique."

data_list = []
data_error_list = []
with open('../data/brands.json', 'r') as file:
    for line in file:
        try:
            data = json.loads(line)
            data_list.append(data)
        except json.JSONDecodeError as e:
            data_error_list.append(line)
            print(f"Error decoding JSON: {e}")
is_oid_unique(data_list)

(True, 'All _id.$oid are unique.')

---------------

2. Design the postgreSQL schema for `brands` table.
   - `brands.brand_code` will be the foreign key referencing `rewards_receipt_items.brand_code`
   - I notic most of time `name` == `brand_code`, the exceiption cases are `brand_code` are null or the difference of upper and lower case letters.

|   | JSON Key     | PostgreSQL Column Name | Data Type   | Key Constraints       | Description       |
|---|--------------|------------------------|-------------|--------------------------|------------------|
| 1 | _id.$oid     | brand_id               |  VARCHAR(24) | PRIMARY KEY    |  brand uuid    |
| 2 | brandCode    | brand_code             | VARCHAR(50)  | FOREIGN KEY    | String that corresponds with the brand column in a partner product file  |
| 3 | barcode      | barcode                | TEXT        |  NOT NULL            |  the barcode on the item   |
| 4 | category     | category               | TEXT        | NULLABLE          | The category name for which the brand sells products in  |
| 5 | categoryCode | category_code          | TEXT        | NULLABLE           | The category code that references a BrandCategory   |
| 6 | cpg.$ref     | cpg_ref                | TEXT        | NOT NULL           | Reference to the CPG ($ref)          |
| 7 | cpg.$id.$oid | cpg_id                 | VARCHAR(24) | NOT NULL            | CPG Identifier ($id.$oid)            |
| 8 |  name        | name                   | TEXT        | NOT NULL     | Brand name    |
| 9 | topBrand    | top_brand              | BOOLEAN     | DEFAULT FALSE   | Boolean indicator for whether the brand should be featured as a 'top brand'  |



------------

3. Create the `brands` table by SQL with corresponding columns and data types.

In [175]:
%%sql
-- delete existing tables if they exist
DROP TABLE IF EXISTS brands CASCADE;
-- create brands table
CREATE TABLE brands (
    brand_id VARCHAR(24) PRIMARY KEY,
    brand_code VARCHAR(50),
    barcode TEXT NOT NULL,
    category TEXT,
    category_code TEXT,
    cpg_ref TEXT NOT NULL,
    cpg_id VARCHAR(24) NOT NULL,
    name TEXT NOT NULL,
    top_brand BOOLEAN DEFAULT FALSE
);

SELECT * FROM brands LIMIT 1;

 * postgresql://postgres:***@155.248.207.242:5432/fetch_db
Done.
Done.
0 rows affected.


[]

--------------------------------

4. Validate the `brands` table schema and ER-Diagram.

<img src = "../images/012.png">

------------------

<br>

#### <a id="3.1.3.3">3.1.3.3 Data ETL (Transfer JSON data to relational database)</a>

<a href="#1.Summary">Click this Link back to Top</a>

1. Create a temporary table `brands_raw` to upload JSON file into remote PostgreSQL database.
2. Use Python to upload the JSON file into the temporary `brands_raw` table.
3. Use SQL to extract data from the temporary `brands_raw` table and load/INSERT data into table `brands`
4. Validate the `brands` table by join it with `rewards_receipt_items` table.

------------------

1. Create a temporary table `brands_raw` to upload JSON file into remote PostgreSQL database.

In [157]:
%%sql
DROP TABLE IF EXISTS brands_raw;
CREATE TABLE brands_raw (
    id SERIAL PRIMARY KEY,
    json_data JSONB
);

SELECT * FROM brands_raw;

 * postgresql://postgres:***@155.248.207.242:5432/fetch_db
Done.
Done.
0 rows affected.


[]

--------------------

2. Use Python to upload the JSON file into the temporary `brands_raw` table.

In [158]:
import psycopg2
import json

# Connect to the remote PostgreSQL database
# [TODO]: refactor the code by using asynchio to handle large datasets
# [TODO]: replace the connection information with .env related variables to hide sensitive information
db_connection = psycopg2.connect(
    host="155.248.207.242",   
    dbname="fetch_db",
    user="postgres",
    password="00wasabi00",
    port="5432"
)
cursor = db_connection.cursor()


data_list = []
data_error_list = []
with open('../data/brands.json', 'r') as file:
    for line in file:
        try:
            data = json.loads(line)
            data_list.append(data)
        except json.JSONDecodeError as e:
            data_error_list.append(line)
            print(f"Error decoding JSON: {e}")

try:
    # Insert each JSON object separately
    insert_query = "INSERT INTO brands_raw (json_data) VALUES (%s)"

    for record in data_list:
        cursor.execute(insert_query, (json.dumps(record),))  # Convert dict to JSONB format
    
    # Commit changes
    db_connection.commit()
    print(f"Successfully inserted {len(data_list)} records into PostgreSQL!")

except Exception as e:
    print(f"Error: {e}")

finally:
    # Close the connection
    cursor.close()
    db_connection.close()

Successfully inserted 1167 records into PostgreSQL!


In [160]:
%%sql
SELECT COUNT(*) FROM brands_raw;

 * postgresql://postgres:***@155.248.207.242:5432/fetch_db
1 rows affected.


[(1167,)]

In [161]:
%%sql
SELECT * FROM brands_raw LIMIT 1;

 * postgresql://postgres:***@155.248.207.242:5432/fetch_db
1 rows affected.


[(1, {'_id': {'$oid': '601ac115be37ce2ead437551'}, 'cpg': {'$id': {'$oid': '601ac114be37ce2ead437550'}, '$ref': 'Cogs'}, 'name': 'test brand @1612366101024', 'barcode': '511111019862', 'category': 'Baking', 'topBrand': False, 'categoryCode': 'BAKING'})]

----------------

3. Use SQL to extract data from the temporary `brands_raw` table and load/INSERT data into table `brands`

In [176]:
%%sql
INSERT INTO brands (
    brand_id, 
    brand_code, 
    barcode, 
    category, 
    category_code, 
    cpg_ref, 
    cpg_id, 
    name,
    top_brand
)
SELECT 
    (json_data ->'_id'->>'$oid')::VARCHAR(24) AS brand_id,
    (json_data ->> 'brandCode') AS brand_code,
    (json_data ->> 'barcode') AS barcode,
    (json_data ->> 'category') AS category,
    (json_data ->> 'categoryCode') AS category_code,
    (json_data -> 'cpg' ->> '$ref')::TEXT AS cpg_ref,
    (json_data -> 'cpg' -> '$id' ->> '$oid')::VARCHAR(24) AS cpg_id,
    (json_data ->> 'name') AS name,
    (json_data ->> 'topBrand')::BOOLEAN AS top_brand
FROM 
    brands_raw;

SELECT * FROM brands LIMIT 2;


 * postgresql://postgres:***@155.248.207.242:5432/fetch_db
1167 rows affected.
2 rows affected.


[('601ac115be37ce2ead437551', None, '511111019862', 'Baking', 'BAKING', 'Cogs', '601ac114be37ce2ead437550', 'test brand @1612366101024', False),
 ('601c5460be37ce2ead43755f', 'STARBUCKS', '511111519928', 'Beverages', 'BEVERAGES', 'Cogs', '5332f5fbe4b03c9a25efd0ba', 'Starbucks', False)]

----------------

4. Validate the `brands` table by join it with `rewards_receipt_items` table by using brand_code as foreign key.

In [179]:
%%sql
SELECT
    COUNT(b.brand_id)
FROM
    brands AS b
    INNER JOIN rewards_receipt_items AS rri ON b.brand_code = rri.brand_code

 * postgresql://postgres:***@155.248.207.242:5432/fetch_db
1 rows affected.


[(635,)]

In [180]:
%%sql
SELECT
    COUNT(b.brand_id)
FROM
    brands AS b
    INNER JOIN rewards_receipt_items AS rri ON b.name = rri.brand_code

 * postgresql://postgres:***@155.248.207.242:5432/fetch_db
1 rows affected.


[(89,)]

-----------------

### <a id="3.1.4">3.1.4 Summary</a>
<a href="#1.Summary">Click this Link back to Top</a>

1. Created four tables, `receipts`, `rewards_receipt_items`, `users`, and `brands`.
2. Column `rewards_receipt_items.receipt_id` is a foreign key that references `receipts.receipt_id`. This is one to many relationship.
3. Column `users.user_id` is a foreign key that references `receipts.user_id`. This is many to many relationship.
4. Column `brands.brand_code` is a foreign key that references `rewards_receipt_items.brand_code`. This is many to many relationship.
5. Table `users` major used to track user login and activities information.
6. Table `brands` major used to track brand's category and top_brand information.

<img src="../images/014.png">

In [181]:
%%sql
SELECT column_name, data_type, is_nullable, column_default
FROM information_schema.columns
WHERE table_name = 'receipts';


 * postgresql://postgres:***@155.248.207.242:5432/fetch_db
14 rows affected.


[('purchase_date', 'timestamp with time zone', 'YES', None),
 ('purchased_item_count', 'integer', 'YES', None),
 ('total_spent', 'numeric', 'YES', None),
 ('finished_date', 'timestamp with time zone', 'YES', None),
 ('create_date', 'timestamp with time zone', 'NO', None),
 ('date_scanned', 'timestamp with time zone', 'NO', None),
 ('bonus_points_earned', 'numeric', 'YES', None),
 ('modify_date', 'timestamp with time zone', 'NO', None),
 ('points_awarded_date', 'timestamp with time zone', 'YES', None),
 ('points_earned', 'numeric', 'YES', None),
 ('bonus_points_earned_reason', 'text', 'YES', None),
 ('rewards_receipt_status', 'character varying', 'NO', None),
 ('user_id', 'character varying', 'NO', None),
 ('receipt_id', 'character varying', 'NO', None)]

-----------------------

## <a id="3.2">3.2 Generate a query that answers a predetermined business question</a>
<a href="#1.Summary">Click this Link back to Top</a>

Write SQL queries against your new structured relational data model that answer at least two of the following bullet points below of your choosing. Commit them to the git repository along with the rest of the exercise.

Note: When creating your data model be mindful of the other requests being made by the business stakeholder. If you can capture more than two bullet points in your model while keeping it clean, efficient, and performant, that benefits you as well as your team.

We use PostgreSQL as our SQL platform

<img src = "../images/13_postgresql-logo.png">

### <a id="3.2.1">3.2.1 What are the top 5 brands by receipts scanned for most recent month?</a>
<a href="#1.Summary">Click this Link back to Top</a>

The information we need:
1. Top 5 brands information come from `rewards_receipt_items.brand_code`. There are total 2600 records in this table which brand_code is not null.
2. The scanned receipts time is coming from table `receipt` and column `date_scanned`. Need to join two tables to get all information.

-----------------------

We only found 6 rows/recods in Feb,2021. There are the most recent records that brand_code data is not empty.

In [ ]:
%%sql
WITH with_latest_month AS(
    -- find the most recent month in the receipts
    SELECT 
        DATE_TRUNC('month', MAX(date_scanned)) AS recent_month
    FROM 
        receipts AS r
        INNER JOIN rewards_receipt_items AS rri ON r.receipt_id = rri.receipt_id
    -- filter out receipts without brand codes
    WHERE
        rri.brand_code IS NOT NULL
),
with_rri AS(
    SELECT
        r.receipt_id,
        r.date_scanned,
        rri.brand_code
    FROM
        rewards_receipt_items AS rri
        INNER JOIN receipts AS r ON rri.receipt_id = r.receipt_id
    WHERE
        -- the month of the receipt match with the most recent month we filtered before
        DATE_TRUNC('month', r.date_scanned) = (SELECT recent_month FROM with_latest_month)
        AND rri.brand_code IS NOT NULL
)
SELECT * FROM with_rri LIMIT 20;

 * postgresql://postgres:***@155.248.207.242:5432/fetch_db
6 rows affected.


[('601890d50a720f05f4000057', datetime.datetime(2021, 2, 1, 23, 37, 57, tzinfo=datetime.timezone.utc), 'BRAND'),
 ('60189cc10a7214ad28000050', datetime.datetime(2021, 2, 2, 0, 28, 49, tzinfo=datetime.timezone.utc), 'BRAND'),
 ('60189cae0a7214ad2800004f', datetime.datetime(2021, 2, 2, 0, 28, 30, tzinfo=datetime.timezone.utc), 'MISSION'),
 ('602454340a720f05a80001ae', datetime.datetime(2021, 2, 10, 21, 46, 28, tzinfo=datetime.timezone.utc), 'VIVA'),
 ('602538b30a7214d8e9000248', datetime.datetime(2021, 2, 11, 14, 1, 23, tzinfo=datetime.timezone.utc), 'BRAND'),
 ('602538740a7214d8e9000246', datetime.datetime(2021, 2, 11, 14, 0, 20, tzinfo=datetime.timezone.utc), 'MISSION')]

-----------------------

- BRNAD
- MISSION
- VIVA
<br>
These three brands are the top 5 brands by receipt scanned date for the most recent month of Feb, 2021

In [244]:
%%sql
WITH with_latest_month AS(
    -- find the most recent month in the receipts
    SELECT 
        DATE_TRUNC('month', MAX(date_scanned)) AS recent_month
    FROM 
        receipts AS r
        INNER JOIN rewards_receipt_items AS rri ON r.receipt_id = rri.receipt_id
    -- filter out receipts without brand codes
    WHERE
        rri.brand_code IS NOT NULL
),
with_rri AS(
    SELECT
        rri.brand_code,
        COUNT(r.receipt_id) AS receipt_count,
        EXTRACT(MONTH FROM r.date_scanned) AS month
    FROM
        rewards_receipt_items AS rri
        INNER JOIN receipts AS r ON rri.receipt_id = r.receipt_id
    WHERE
        DATE_TRUNC('month', r.date_scanned) = (SELECT recent_month FROM with_latest_month)
        AND rri.brand_code IS NOT NULL
    GROUP BY
        rri.brand_code,
        month
    ORDER BY
        receipt_count DESC
)
SELECT * FROM with_rri;

 * postgresql://postgres:***@155.248.207.242:5432/fetch_db
3 rows affected.


[('BRAND', 3, Decimal('2')),
 ('MISSION', 2, Decimal('2')),
 ('VIVA', 1, Decimal('2'))]

The top five brands based on receipt scans for the most recent month are as follows.
- BRAND (3)
- MISSION (2)
- VIVA (1)

-----------------

But i think the previous month Jan,2021 have more valuable data then Feb, 2021.
So we track back to one more month Jan,2021 and get the data.


In [243]:
%%sql
WITH with_latest_month AS(
    -- find the most recent month in the receipts
    SELECT
        -- track back one more month from the recent non-null month and set it as recent_month
        DATE_TRUNC('month', MAX(date_scanned)) - INTERVAL '1 month' AS recent_month
    FROM 
        receipts AS r
        INNER JOIN rewards_receipt_items AS rri ON r.receipt_id = rri.receipt_id
    -- filter out receipts without brand codes
    WHERE
        rri.brand_code IS NOT NULL
),
with_rri AS(
    SELECT
        rri.brand_code,
        COUNT(r.receipt_id) AS receipt_count,
        EXTRACT(month FROM r.date_scanned) AS month
    FROM
        rewards_receipt_items AS rri
        INNER JOIN receipts AS r ON rri.receipt_id = r.receipt_id
    WHERE
        DATE_TRUNC('month', r.date_scanned) = (SELECT recent_month FROM with_latest_month)
        AND rri.brand_code IS NOT NULL
    -- categorize by brand_code
    GROUP BY
        rri.brand_code,
        month
    -- rand the order of brand count
    ORDER BY
        receipt_count DESC
    -- filter out the top 5 brand codes
    LIMIT
        5
)
SELECT * FROM with_rri;

 * postgresql://postgres:***@155.248.207.242:5432/fetch_db
5 rows affected.


[('HY-VEE', 291, Decimal('1')),
 ('BEN AND JERRYS', 180, Decimal('1')),
 ('PEPSI', 93, Decimal('1')),
 ('KROGER', 89, Decimal('1')),
 ('KLEENEX', 88, Decimal('1'))]

The top five brands based on receipt scans for the month preceding the most recent month are as follows.
<br>
**HY_VEE** was scanned the most number of times with a total of 291 receipts. 
- HY-VEE (291)
- BEN AND JERRYS (180)
- PEPSI (165)
- KROGER (89)
- KLEENEX (88)

------------------------

<br>

### <a id="3.2.2">3.2.2 How does the ranking of the top 5 brands by receipts scanned for the recent month compare to the ranking for the previous month?</a>
<a href="#1.Summary">Click this Link back to Top</a>

1. The first approrach is to UNITON my previous two month together. But the code is two large.
2. The second approach is to use the same WITH() and full out join to find the compare result.
3. COMBINE year and month to create a new column `year_month`
4. Add ranking column to prepare the data trend.


--------------

1. The first approrach is to UNITON my previous two month together. But the code is two large.

In [274]:
%%sql
WITH with_two_months AS(
    -- find the most recent month in the receipts
    SELECT 
        DATE_TRUNC('month', MAX(date_scanned)) AS recent_month,
        DATE_TRUNC('month', MAX(date_scanned)) - INTERVAL '1 month' AS previous_month
    FROM 
        receipts AS r
        INNER JOIN rewards_receipt_items AS rri ON r.receipt_id = rri.receipt_id
    -- filter out receipts without brand codes
    WHERE
        rri.brand_code IS NOT NULL
),
with_recent_month_ranking AS(
    SELECT
        rri.brand_code,
        COUNT(r.receipt_id) AS receipt_count,
        -- extract the year and month from the date_scanned column and concatenate them into a string
        TO_CHAR(r.date_scanned, 'YYYY-MM') AS year_month
        -- calculate the rank of each brand based on the number of receipts in the most recent month
        -- ROW_NUMBER() OVER (ORDER BY COUNT(r.receipt_id) DESC) AS rank
    FROM
        rewards_receipt_items AS rri
        INNER JOIN receipts AS r ON rri.receipt_id = r.receipt_id
    WHERE
        DATE_TRUNC('month', r.date_scanned) = (SELECT recent_month FROM with_two_months)
        AND rri.brand_code IS NOT NULL
    GROUP BY
        rri.brand_code,
        year_month
    ORDER BY
        receipt_count DESC
    LIMIT 5
),
with_previous_month_ranking AS(
    SELECT
        rri.brand_code,
        COUNT(r.receipt_id) AS receipt_count,
        -- extract the year and month from the date_scanned column and concatenate them into a string
        TO_CHAR(r.date_scanned, 'YYYY-MM') AS year_month
        -- calculate the rank of each brand based on the number of receipts in the most recent month
        -- ROW_NUMBER() OVER (ORDER BY COUNT(r.receipt_id) DESC) AS rank
    FROM
        rewards_receipt_items AS rri
        INNER JOIN receipts AS r ON rri.receipt_id = r.receipt_id
    WHERE
        DATE_TRUNC('month', r.date_scanned) = (SELECT previous_month FROM with_two_months)
        AND rri.brand_code IS NOT NULL
    GROUP BY
        rri.brand_code,
        year_month
    ORDER BY
        receipt_count DESC
    LIMIT 5
),
with_union AS(
    SELECT * FROM with_recent_month_ranking
    UNION ALL
    SELECT * FROM with_previous_month_ranking
)
SELECT * FROM with_union


 * postgresql://postgres:***@155.248.207.242:5432/fetch_db
8 rows affected.


[('BRAND', 3, '2021-02'),
 ('MISSION', 2, '2021-02'),
 ('VIVA', 1, '2021-02'),
 ('HY-VEE', 291, '2021-01'),
 ('BEN AND JERRYS', 180, '2021-01'),
 ('PEPSI', 93, '2021-01'),
 ('KROGER', 89, '2021-01'),
 ('KLEENEX', 88, '2021-01')]

From previous table, we know:
1. The recent month is Feb, 2021 and the previous month is Jan, 2021.
2. In Feb,2021, only 6 records and `BRAND` is the most popular brand.
3. In Jan,2021, `HY-VEE` is the most popular brand.
4. No Jan.2021 top 5 brands shows again in Feb,2021. 

---------------

2. The second approach is to use the same WITH() and full out join to discover the compare ranking result.

In [272]:
%%sql
WITH with_two_months AS(
    -- find the most recent month in the receipts
    SELECT 
        DATE_TRUNC('month', MAX(date_scanned)) AS recent_month,
        DATE_TRUNC('month', MAX(date_scanned)) - INTERVAL '1 month' AS previous_month
    FROM 
        receipts AS r
        INNER JOIN rewards_receipt_items AS rri ON r.receipt_id = rri.receipt_id
    -- filter out receipts without brand codes
    WHERE
        rri.brand_code IS NOT NULL
),
with_recent_month_ranking AS(
    SELECT
        rri.brand_code,
        COUNT(r.receipt_id) AS receipt_count,
        -- extract the year and month from the date_scanned column and concatenate them into a string
        TO_CHAR(r.date_scanned, 'YYYY-MM') AS year_month,
        -- calculate the rank of each brand based on the number of receipts in the most recent month
        ROW_NUMBER() OVER (ORDER BY COUNT(r.receipt_id) DESC) AS rank
    FROM
        rewards_receipt_items AS rri
        INNER JOIN receipts AS r ON rri.receipt_id = r.receipt_id
    WHERE
        DATE_TRUNC('month', r.date_scanned) = (SELECT recent_month FROM with_two_months)
        AND rri.brand_code IS NOT NULL
    GROUP BY
        rri.brand_code,
        year_month
    LIMIT 5
),
with_previous_month_ranking AS(
    SELECT
        rri.brand_code,
        COUNT(r.receipt_id) AS receipt_count,
        -- extract the year and month from the date_scanned column and concatenate them into a string
        TO_CHAR(r.date_scanned, 'YYYY-MM') AS year_month,
        -- calculate the rank of each brand based on the number of receipts in the most recent month
        ROW_NUMBER() OVER (ORDER BY COUNT(r.receipt_id) DESC) AS rank
    FROM
        rewards_receipt_items AS rri
        INNER JOIN receipts AS r ON rri.receipt_id = r.receipt_id
    WHERE
        DATE_TRUNC('month', r.date_scanned) = (SELECT previous_month FROM with_two_months)
        AND rri.brand_code IS NOT NULL
    GROUP BY
        rri.brand_code,
        year_month
    LIMIT 5
)
SELECT
    COALESCE(rec.brand_code, pre.brand_code) AS brand_code,
    rec.year_month AS recent_month,
    rec.receipt_count AS recent_month_count,
    rec.rank AS recent_month_rank,
    --CONCAT(rec.rank, '(', rec.receipt_count, ')') AS recent_month_rank_count,
    pre.year_month AS previous_month,
    pre.receipt_count AS prevous_month_count,
    pre.rank AS previous_month_rank,
    CASE
        WHEN rec.rank IS NULL THEN 'Dropped from TOP 5'
        WHEN pre.rank IS NULL THEN 'New to TOP 5'
        WHEN rec.rank < pre.rank THEN '↓ Declined'
        WHEN rec.rank > pre.rank THEN '↑ Improved'
        ELSE '➔ No change'
    END AS trend
FROM
    with_recent_month_ranking AS rec
    FULL OUTER JOIN with_previous_month_ranking AS pre ON rec.brand_code = pre.brand_code
ORDER BY
    COALESCE(rec.rank, pre.rank) ASC


 * postgresql://postgres:***@155.248.207.242:5432/fetch_db
8 rows affected.


[('BRAND', '2021-02', 3, 1, None, None, None, 'New to TOP 5'),
 ('HY-VEE', None, None, None, '2021-01', 291, 1, 'Dropped from TOP 5'),
 ('MISSION', '2021-02', 2, 2, None, None, None, 'New to TOP 5'),
 ('BEN AND JERRYS', None, None, None, '2021-01', 180, 2, 'Dropped from TOP 5'),
 ('PEPSI', None, None, None, '2021-01', 93, 3, 'Dropped from TOP 5'),
 ('VIVA', '2021-02', 1, 3, None, None, None, 'New to TOP 5'),
 ('KROGER', None, None, None, '2021-01', 89, 4, 'Dropped from TOP 5'),
 ('KLEENEX', None, None, None, '2021-01', 88, 5, 'Dropped from TOP 5')]

From previous table, we know:
1. The recent_month(2021-02) brands are: `BRNAD`, `MISSION`, AND `VIVA`. But they didn't appear in the previous month(2021-01).
2. On the other hand, the previous_month(2021-01) brands also didn't appear in the recent month(2021-02).
3. The trend don't have any siginificent consisitency.

-------------------------

<br>

### <a id="3.2.3">3.2.3 When considering average spend from receipts with 'rewardsReceiptStatus’ of ‘Accepted’ or ‘Rejected’, which is greater?</a>
<a href="#1.Summary">Click this Link back to Top</a>

The first insights are:
1. We don't have **Accepted** in status.
2. If we consider **FINISHED** as **Accepted**, the conslusion is **Finished** spend is higher then **Rejected** spend

In [280]:
%%sql
SELECT
    COUNT(receipt_id),
    rewards_receipt_status,
    ROUND(AVG(total_spent::float)) AS avg_total_spent
FROM
    receipts
GROUP BY
    rewards_receipt_status
LIMIT 10;


 * postgresql://postgres:***@155.248.207.242:5432/fetch_db
5 rows affected.


[(71, 'REJECTED', 23.0),
 (434, 'SUBMITTED', None),
 (46, 'FLAGGED', 180.0),
 (50, 'PENDING', 28.0),
 (518, 'FINISHED', 81.0)]

------------------------------

2. If we consider all status except **Rejected** as **Accepted**, then **Accepted** average spend $84 is higher then **Rejected** average spend $23.

In [283]:
%%sql
SELECT
    COUNT(receipt_id),
    CASE
        WHEN rewards_receipt_status = 'REJECTED' THEN 'Rejected'
        ELSE 'Accepted'
    END AS receipt_status,
    ROUND(AVG(total_spent::float)) AS avg_total_spent
FROM
    receipts
GROUP BY
    receipt_status

 * postgresql://postgres:***@155.248.207.242:5432/fetch_db
2 rows affected.


[(71, 'Rejected', 23.0), (1048, 'Accepted', 84.0)]

------------------------

<br>

### <a id="3.2.4">3.2.4 When considering total number of items purchased from receipts with 'rewardsReceiptStatus’ of ‘Accepted’ or ‘Rejected’, which is greater?</a>
<a href="#1.Summary">Click this Link back to Top</a>

1. My first thinking is where should the infomation **total number of items purchsed** coming from? Should it be from **receipt.purchased_item** or should it be from **COUNT(rewards_receipt_items.item_id)**
2. The following SQL query told me that these two metrics are not equal. 25% of data are not equal. Only 75% of the data are equal.


In [330]:
%%sql
WITH with_list AS(
    SELECT
        r.receipt_id,
        COUNT(rri.item_id) AS list_count,
        r.purchased_item_count
    FROM
        rewards_receipt_items AS rri
        INNER JOIN receipts AS r ON rri.receipt_id = r.receipt_id
    WHERE
        r.purchased_item_count > 0
    GROUP BY
        -- for each receipt, count the number of unique items purchased and also the total number of rewardsReceipItemsList
        r.receipt_id
),
with_equal AS(
    SELECT
        COUNT(*),
        CASE
            WHEN list_count = purchased_item_count THEN 'equal'
            WHEN list_count > purchased_item_count THEN 'list over purchased_count'
            ELSE 'list under purchased_count'
        END AS compare_result
    FROM
        with_list
    GROUP BY
        compare_result
)
SELECT * FROM with_equal;

 * postgresql://postgres:***@155.248.207.242:5432/fetch_db
3 rows affected.


[(8, 'list over purchased_count'),
 (466, 'equal'),
 (146, 'list under purchased_count')]

-----------------

In order to answer this question, we need to make some assumptions:
1. **total number of items purchsed** will use `receipts.purchased_item_count` column.
2. I will use **FINISHED** status to represent **Accepted** status.

In [300]:
%%sql
SELECT
    rewards_receipt_status,
    SUM(purchased_item_count) AS total_items_purchased
FROM
    receipts
WHERE  
    rewards_receipt_status IN ('FINISHED', 'REJECTED')
    AND purchased_item_count IS NOT NULL
GROUP BY
    rewards_receipt_status
ORDER BY
    total_items_purchased


 * postgresql://postgres:***@155.248.207.242:5432/fetch_db
2 rows affected.


[('REJECTED', 173), ('FINISHED', 8184)]

Insights:
1. 97.79% purchased items were **Accepted** by Fetch Rewards system.

-------------------------

<br>

### <a id="3.2.5">3.2.5 Which brand has the most spend among users who were created within the past 6 months?</a>
<a href="#1.Summary">Click this Link back to Top</a>

Anaylsis:
1. The information of users who create within the past 6 months comes from the `users` table.
2. brand information comes from `rewards_receipt_items` table.
3. spend information may reference to `receipts.total_spent` column.
4. Assume that the questions was asked when data was collected. Therefore, we can understand the time period of data should the last records and then track back to another 6 month. Finally, the time period of this question will be Aug, 2020 to Feb,2021.

In [ ]:
%%sql
WITH with_time AS(
    SELECT
        -- the most recent created_date records in users table
        DATE_TRUNC('month', MAX(created_date)) AS last_created_month,
        DATE_TRUNC('month', MAX(created_date)) - INTERVAL '6 month' AS six_month_ago
    FROM
        users
),
with_users AS(
    -- identify users created in the most recent 6 months
    SELECT
        record_id,
        user_id,
        created_date
    FROM
        users
    WHERE
        created_date >= (SELECT six_month_ago FROM with_time)
),
with_receipts AS(
    -- get the brand_code info from reward_receipt_items table and join with receipts table
    SELECT
        -- r.receipt_id,
        rri.brand_code,
        SUM(r.total_spent) AS sum_total_spend
    FROM
        receipts AS r
        INNER JOIN rewards_receipt_items AS rri ON r.receipt_id = rri.receipt_id
        INNER JOIN with_users AS u ON r.user_id = u.user_id
    WHERE
        rri.brand_code <> 'None'
    GROUP BY
        rri.brand_code
    ORDER BY
        sum_total_spend DESC
    LIMIT 5
)
SELECT * FROM with_receipts;

 * postgresql://postgres:***@155.248.207.242:5432/fetch_db
5 rows affected.


[('HY-VEE', Decimal('373558.41')),
 ('BEN AND JERRYS', Decimal('227078.39')),
 ('KROGER', Decimal('222538.59')),
 ('PEPSI', Decimal('119703.34')),
 ('KLEENEX', Decimal('96896.05'))]

Summary:
1. For the Users who created within past 6 months (2020-08 to 2021-02 ), the most spend brand is **HY-VEE**

--------------

<br>

### <a id="3.2.6">3.2.6 Which brand has the most transactions among users who were created within the past 6 months?</a>
<a href="#1.Summary">Click this Link back to Top</a>

Analysis:
1. Users within the past 6 months will be extracted from 'users' table.
2. Since I didn't find one column that contains **transaction**, I will assume per receipt per transaction, which identified by `receipt` tables.
3. Assume that the questions was asked when data was collected. Therefore, we can understand the time period of data should the last records and then track back to another 6 month. Finally, the time period of this question will be Aug, 2020 to Feb,2021.

In [327]:
%%sql
WITH with_time AS(
    SELECT
        -- the most recent created_date records in users table
        DATE_TRUNC('month', MAX(created_date)) AS last_created_month,
        DATE_TRUNC('month', MAX(created_date)) - INTERVAL '6 month' AS six_month_ago
    FROM
        users
),
with_users AS(
    -- identify users created in the most recent 6 months
    SELECT
        record_id,
        user_id,
        created_date
    FROM
        users
    WHERE
        created_date >= (SELECT six_month_ago FROM with_time)
),
with_join AS (
    SELECT
        rri.brand_code,
        COUNT(r.receipt_id) AS total_transactions
    FROM
        receipts AS r
        INNER JOIN with_users AS u ON r.user_id = u.user_id
        INNER JOIN rewards_receipt_items AS rri ON r.receipt_id = rri.receipt_id
    WHERE
        rri.brand_code <> 'None'
    GROUP BY
        rri.brand_code
    ORDER BY
        total_transactions DESC
)
SELECT * FROM with_join LIMIT 5;

 * postgresql://postgres:***@155.248.207.242:5432/fetch_db
5 rows affected.


[('HY-VEE', 738),
 ('BEN AND JERRYS', 130),
 ('PEPSI', 118),
 ('KLEENEX', 115),
 ('DOLE', 89)]

Summary:
1. **HY-VEE** has the most transactions for the users who created in the past 6 months.
2. Compare to previous result, **HY-VEE** also have most spend and most transactions.

------------------------

<br>

## <a id="3.3">3.3 Evaluate Data Quality Issues in the Data Provided</a>
<a href="#1.Summary">Click this Link back to Top</a>

Using the programming language of your choice (SQL, Python, R, Bash, etc...) identify as many data quality issues as you can. We are not expecting a full blown review of all the data provided, but instead want to know how you explore and evaluate data of questionable provenance.

1. In `brands` table, we find `brand_code` sometime have uppercase letter version comparing with `name` column. So they delivery some infomation but in different format. We should convert them into same format.
2. The data missing probably in each tables.
3. The data imbalance indicate the data extraction process is not accurate. For example, 90% of records in `users` table is coming from **Jan, 2021**.  

Data Clean
1. extract *.gz data and get the original JSON file
2. This is typical front-end data and maybe coming form MongoDB.
3. Create *_raw table for the original JSON data.
4. [Optional] If the PostgreSQL database is in local, I can directly use SQL or psql to import the raw json data
    ```SQL
    # use the COPY command to import the JSON file into receipt_raw table in the PostgreSQL database.
    %%sql
    COPY receipts_raw(data)
    FROM '../data/receipts.json'
    WITH (FROMT json);

    ```
5. I try to use SQL to import this JSON file. But the JSON file have nested JSON structure and some format is not standard. Considering data clean process should be part of automation data pipeline, I choose to use python to complete the data clean and data import process. In the future, we can use Airflow or AWS auto to complete this data pipeline.
6. Since my PostgreSQL database is not in local. So I choose use python script to import the local JSON file into remote database

Future work
1. Add .env to hide sensitive information, such as ip address, username, password, etc.

In [ ]:
%%sql
CREATE OR REPLACE FUNCTION insert_receipts_with_logging()
RETURNS VOID AS $$
DECLARE
    json_record JSONB;
    error_context TEXT;
BEGIN
    FOR json_record IN SELECT data FROM receipts_raw LOOP
        BEGIN
            INSERT INTO receipts (
                receipt_id,
                create_date,
                date_scanned,
                modify_date,
                rewards_receipt_status,
                user_id,
                bonus_points_earned,
                bonus_points_earned_reason,
                points_awarded_date,
                points_earned,
                purchase_date,
                purchased_item_count,
                total_spent,
                finished_date
            )
            VALUES (
                (json_record -> '_id' ->> '$oid')::VARCHAR(24),
                (json_record -> 'createDate' ->> '$date')::TIMESTAMPTZ,
                (json_record -> 'dateScanned' ->> '$date')::TIMESTAMPTZ,
                (json_record -> 'modifyDate' ->> '$date')::TIMESTAMPTZ,
                (json_record ->> 'rewardsReceiptStatus')::VARCHAR(50),
                (json_record ->> 'userId')::VARCHAR(24),
                (json_record ->> 'bonusPointsEarned')::DECIMAL(10,2),
                (json_record ->> 'bonusPointsEarnedReason')::TEXT,
                (json_record -> 'pointsAwardedDate' ->> '$date')::TIMESTAMPTZ,
                (json_record ->> 'pointsEarned')::DECIMAL(10,2),
                (json_record -> 'purchaseDate' ->> '$date')::TIMESTAMPTZ,
                (json_record ->> 'purchasedItemCount')::INTEGER,
                (json_record ->> 'totalSpent')::DECIMAL(10,2),
                (json_record -> 'finishedDate' ->> '$date')::TIMESTAMPTZ
            );

        EXCEPTION
            WHEN OTHERS THEN
                GET STACKED DIAGNOSTICS error_context = PG_EXCEPTION_CONTEXT;
                INSERT INTO error_log (
                    receipt_id,
                    error_message,
                    json_data
                )
                VALUES (
                    (json_record -> '_id' ->> '$oid')::VARCHAR(24),
                    SQLERRM || ' | Context: ' || error_context,
                    json_record
                );
        END;
    END LOOP;
END;
$$ LANGUAGE plpgsql;